### PR TITLE
Implement type definitions generation via JSDoc comments

### LINKED ISSUE #1888: TypeScript type definitions for PhotoSwipe v5
There are no type definitions for v5 currently available.

The package with type definitions [@types/photoswipe](https://www.npmjs.com/package/@types/photoswipe) for v4 currently has about 30k weekly downloads.

image

Meanwhile, the [photoswipe](https://www.npmjs.com/package/photoswipe) package itself has about 100k weekly downloads on NPM.

image

It means that almost a third of users need it.

It would be great to have built-in type definitions for v5, right in the PhotoSwipe package.

I see some ways to achieve this:

1) Make the `.d.ts` file manually in this repo.
Pros: it's easy; no need to modify an existing codebase.
Cons: it's hard to keep it in sync; no typechecking for the PhotoSwipe source.

2) Rewrite the PhotoSwipe source code to TypeScript.
Pros: strongly-typed codebase; automatically generated type definitions.
Cons: a lot of work; more complex build stage; inevitable use of TS compiler.

3) Use of a hybrid approach with TypeScript and JSDoc
Pros: easy to implement; automatically generated type definitions; somewhat strongly-typed codebase;
Cons: need to document all the sources with JSDoc 🤔

https://www.typescriptlang.org/docs/handbook/intro-to-js-ts.html
https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html
https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html
https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html
https://github.com/webpack/webpack/blob/main/lib/AsyncDependenciesBlock.js

I think that the third approach is the best option for this project. For example, webpack went this way successfully and it seems like this is a good working solution.

I can do a PR for it 😉

### PR DESCRIPTION
**Pull request by irudoy**
This PR introduces type definitions for the PhotoSwipe. It was a bit harder to do and took much more time than I first assumed. Not everything went as planned, because of some JSDoc TS limitations, but it's finally here!

### What has been done?
- The entire codebase was covered with type definitions.
- Set up `.d.ts` files generation to `dist/types` dir, in `build-and-minify-js` npm script.
- Entry points for types were referenced in the `package.json` in the [actual way](https://github.com/microsoft/TypeScript/issues/33079). Also, a fallback for older TS versions was provided.
- The JS code was modified slightly:
  - Added some class properties.
  - Provided strings instead of numbers or cast them where it's expected.
  - Property lookup using `in` operator in some places, for type guards.
- Some minor chores that affect the entire project:
  - `package-lock.json` updated from `v1` to `v2` (it was done automatically, by running `npm install` with npm v8.5.5).
  - Added `.editorconfig` [file](https://editorconfig.org/) to maintain consistency across different configuration files.
  - Slightly increased value for `max-len`eslint rule (only for comments).
  - Added `.eslintignore` file.

### How will these changes improve UX?
- Library users can use it in TypeScript projects.
- In non-TS projects, users will get intellisense to work in their editors (code completion, type hints, etc.).

### How will these changes improve DX?
- Almost complete type-checking for the entire codebase, all functions, classes, events and filters.
- It will become easier to understand how the code works, what arguments are expected, etc.

### What can be fixed in the future?
- Resolve `// @ts-expect-error` comments.
- Exclude build artifacts from git.
- Implement display of the comments for events/filters in the editors.
- Rewrite all in TS maybe? It's supposed to be pretty easy from this point.
- Auto-generate docs from type definitions/comments?
- Try to differentiate PhotoSwipe and PhotoSwipeLightbox options, if it makes sense.
- Infer type for `instance.options.dataSource` from provided `dataSource` option.
- Define type for `onInit(el)` of `registerElement` based on provided `tagName`.
- Add ability to describe `SlideData` additional props more strictly from the outside somehow.
- Add description to all exposed options/methods.
- Add some tests for types.

I excluded build artifacts from this PR for an easier review process. You can see it here: https://github.com/irudoy/PhotoSwipe/tree/types-built/dist/types

Fixes #1888

### CODE REVIEW

**Comment by dimsemenov** [comment]
Thanks for the tremendous work. I'll take a few days to review and test the PR.

**Comment by irudoy** [comment]
@dimsemenov nice! 

Just tested all examples from the docs, and it mostly works well, instead of a few minor issues that I have fixed already. Also found a few places that would potentially be nice to improve in the future (described in the PR body).

Sharing my test snippet: https://gist.github.com/irudoy/42059d65ec47c507fca803441afade08 Maybe it will be useful for testing.

File: src/js/photoswipe.js [hunk 1]
```diff
@@ -123,6 +330,7 @@ class PhotoSwipe extends PhotoSwipeBase {
     this.offset.y = window.pageYOffset;
 
     this._initialItemData = this.getItemData(this.currIndex);
+    // @ts-expect-error Expected 1-2 arguments, but got 4 ¯\_(ツ)_/¯
     this.dispatch('gettingData', this.currIndex, this._initialItemData, true);
```
Comment (irudoy) [inline src/js/photoswipe.js:334]: This should be fixed maybe?

File: src/js/slide/loader.js [hunk 2]
```diff
@@ -25,6 +31,7 @@ export function lazyLoadData(itemData, instance, index) {
 
   // We need to know dimensions of the image to preload it,
   // as it might use srcset and we need to define sizes
+  // @ts-expect-error should provide pswp instance?
   const viewportSize = instance.viewportSize || getViewportSize(options);
```
Comment (irudoy) [inline src/js/slide/loader.js:35]:
```suggestion
  const viewportSize = instance.viewportSize || getViewportSize(options, instance);
```

File: src/js/slide/slide.js [hunk 3]
```diff
@@ -83,6 +112,7 @@ class Slide {
 
     this.calculateSize();
 
+    // @ts-expect-error: should be `this.container.style.transformOrigin`?
     this.container.transformOrigin = '0 0';
```
Comment (irudoy) [inline src/js/slide/slide.js:116]:
```suggestion
    this.container.style.transformOrigin = '0 0';
```

File: src/js/photoswipe.js (hunk 1 above)
Comment (dimsemenov) [reply src/js/photoswipe.js:348]: yea, it's a typo from a legacy event, will fix it myself. thanks!

File: src/js/ui/ui-element.js [hunk 4]
```diff
@@ -37,21 +75,28 @@ function addElementHTML(htmlData) {
...
+  /**
+   * @param {PhotoSwipe} pswp
+   * @param {UIElementData} data
+   */
   constructor(pswp, data) {
     const name = data.name || data.className;
     let elementHTML = data.html;
 
+    // @ts-expect-error lookup only by `data.name` maybe?
     if (pswp.options[name] === false) {
```
[the remark this answers is not in the document]
Comment (dimsemenov) [reply src/js/ui/ui-element.js:87]: Don't think we can do that, since some people rely on className as an option name, and it'll break the code.

**Comment by dimsemenov** [comment]
Approved 👍 

I'll add some discussed changes myself afterward. Thanks again for the work, hope my code wasn't too hard to read. I might rewrite it in TS in future, but as you see it's a lot of work...

File: src/js/ui/ui-element.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (irudoy) [reply src/js/ui/ui-element.js:87]: Wow, doesn't seen that in the docs

**Comment by irudoy** [comment]
Thanks for approving. No, it was far easier to read in comparison to the v4 source :) Nice work, although!

File: src/js/ui/ui-element.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (dimsemenov) [reply src/js/ui/ui-element.js:87]: yea, it was in an old API, maybe I'll patch it in the major release.

**Comment by BilouMaster** [comment]
You guys are marvelous, thanks for the hard work, I can't wait to update my integration of pswp! 🤘🔥

Obtenir Outlook pour Android<https://aka.ms/AAb9ysg>
________________________________

**Pull request merged**